kafka-client/src/main/java/dev/responsive/kafka/api/ResponsiveGlobalConsumer.java

```diff
@@ -0,0 +1,398 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.api;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerGroupMetadata;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.clients.consumer.OffsetCommitCallback;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * The {@code ResponsiveGlobalConsumer} is a proxy {@link KafkaConsumer} that
+ * allows the Responsive code to use consumer groups for populating a
+ * {@link org.apache.kafka.streams.kstream.GlobalKTable}. It does four things:
+ *
+ * <ol>
+ *   <li>It intercepts calls to {@link #assign(Collection)} and instead translates
+ *   them to {@link #subscribe(Collection)}</li>
+ *
+ *   <li>It ignores calls to {@link #seek(TopicPartition, long)} so that it
+ *   can properly use the committed offsets.</li>
+ *
+ *   <li>It allows looking up {@link #position(TopicPartition)} for partitions
+ *   that this consumer does not own by proxying the Admin for those partitions.</li>
+ *
+ *   <li>It returns {@link ConsumerRecords} from {@link #poll(Duration)} that will
+ *   return all records when calling {@link ConsumerRecords#records(TopicPartition)}
+ *   instead of only the requested partition. This is necessary because of the way that
+ *   the {@link org.apache.kafka.streams.processor.internals.GlobalStateManagerImpl}
+ *   handles restore. Specifically, it will loop partition-by-partition and update
+ *   the store for those partitions. The issue with this is that if we use {@code subscribe}
+ *   in place of {@code assign} we may get events from partitions that were not
+ *   specified in the {@code assign} call. These records will be dropped.</li>
+ *
+ *   <li>It commits before every call to {@link #poll(Duration)} because
```

Comment:
Oof...this could put a LOT of load on the brokers 😬

Reply:
yeah :/ I was actually specifically going to ping you on this one and see if you had other suggestions.